### PR TITLE
VAE

### PR DESCRIPTION
**Pull request by vitskvara**

### CODE REVIEW

File: scripts/evaluate_performance.jl [hunk 1]
```diff
@@ -42,7 +43,7 @@ function query_stats(target::String)
 			@info "$target not compatible"
 		end
 	else
-		query_stats.(joinpath.(target, readdir(target)))
+		query_stats.(joinpath.(target, filter(x->!(occursin("model_", string(x))), readdir(target))))
```
Comment (janfrancu) [inline scripts/evaluate_performance.jl:46]: I have noticed this earlier, thanks for including it in the PR.

File: scripts/experiments_tabular/vae.jl [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+	sample_params()
+
+Should return a named tuple that contains a sample of model parameters.
+"""
+function sample_params()
+	par_vec = (2 .^(3:8), 2 .^(4:9), 10f0 .^(-4:-3), 2 .^ (5:7), ["relu", "swish", "tanh"], 3:4, 1:Int(1e8))
+	argnames = (:zdim, :hdim, :lr, :batchsize, :activation, :nlayers, :init_seed)
+	parameters = (;zip(argnames, map(x->sample(x, 1)[1], par_vec))...)
+	# ensure that zdim < hdim
+	while parameters.zdim >= parameters.hdim
```
Comment (janfrancu) [inline scripts/experiments_tabular/vae.jl:43]: What about putting this into `edit_params`, where you can divide by 2 until the condition is satisfied?

File: scripts/experiments_tabular/vae.jl [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+Each element of the return vector contains a specific anomaly score function - there can be multiple for each trained model.
+Final parameters is a named tuple of names and parameter values that are used for creation of the savefile name.
+"""
+function fit(data, parameters)
+	# construct model - constructor should only accept kwargs
+	model = GenerativeAD.Models.vae_constructor(;idim=size(data[1][1],1), parameters...)
+
+	# fit train data
+	try
+		global info, fit_t, _, _, _ = @timed fit!(model, data; max_train_time=82800,
```
Comment (janfrancu) [inline scripts/experiments_tabular/vae.jl]: I realized yesterday, that the maximum training time of 23 hours does not take into account, that this script runs 5 trainings. I don't know if we should limit the training to 23/5 hours or just leave it be.

File: scripts/experiments_tabular/vae.jl [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+	# now return the different scoring functions
+	training_info, [
+		(x -> GenerativeAD.Models.reconstruction_score(info.model, x), merge(parameters, (score = "reconstruction",))),
+		(x -> GenerativeAD.Models.latent_score(info.model, x), merge(parameters, (score = "latent",))),
+		]
+end
+function GenerativeAD.edit_params(data, parameters)
+	idim = size(data[1][1],1)
+	# put the largest possible zdim where zdim < idim, the model tends to converge poorly if the latent dim is larger than idim
+	if parameters.zdim >= idim
```
Comment (janfrancu) [inline scripts/experiments_tabular/vae.jl:87]: Is this duplicate of the code in `sample_params`?

File: src/models/vae.jl [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+	(nlayers < 2) ? error("Less than 3 layers are not supported") : nothing
+	# function from string
+	act = eval(Meta.parse(activation))
+
+	# if seed is given, set it
+	(init_seed != nothing) ? Random.seed!(init_seed) : nothing
+	
+	# construct the model
+	# encoder - diagonal covariance
+	encoder_map = Chain(
```
Comment (janfrancu) [inline src/models/vae.jl:29]: I see that you did not use the `build_mlp` methods. Is there something missing in their implementation?

Comment (janfrancu) [reply src/models/vae.jl:29]:
```julia
decoder_map = Chain(
		build_mlp(idim, hdim, hdim, nlayers-1, activation=activation, lastlayer="linear")...,
		ConditionalDists.SplitLayer(hdim, [zdim, zdim], [identity, safe_softplus])
		)
```
Should be equivalent, but I have not checked it.

File: src/models/vae.jl [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+"""
+	loss(model::GenerativeModels.VAE, x[, batchsize])
+
+Negative ELBO for training of a VAE model.
+"""
+loss(model::GenerativeModels.VAE, x) = -elbo(model, x)
+# version of loss for large datasets where
+loss(model::GenerativeModels.VAE, x, batchsize::Int) = 
+	mean(map(y->loss(model,y), Flux.Data.DataLoader(x, batchsize=batchsize)))
```
Comment (janfrancu) [inline src/models/vae.jl:57]: Nice use of dispatch.

File: src/models/vae.jl [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+		else # else stop if the model has not improved for `patience` iterations
+			_patience -= 1
+			if _patience == 0
+				@info "Stopped training after $(i) iterations."
+				break
+			end
+		end
+		i += 1
+	end
+	# again, this is not optimal, the model should be passed by reference and only teh reference should be edited
```
Comment (janfrancu) [inline src/models/vae.jl]: spelling *the

File: src/models/vae.jl [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+		batch_loss = 0f0
+		gs = gradient(() -> begin 
+			batch_loss = loss(tr_model,batch)
+		end, ps)
+	 	Flux.update!(opt, ps, gs)
+
+		# validation/early stopping
+		val_loss = (val_N > 5000) ? loss(tr_model, val_x, 256) : loss(tr_model, val_x)
+		
+		(i%check_interval == 0) ? (@info "$i - loss: $(batch_loss) (batch) | $(val_loss) (validation)") : nothing
```
Comment (janfrancu) [inline src/models/vae.jl:90]: Is `check_interval` is used only for printing or did you intend to check also the validation loss also less often? I'm asking just if it makes sense to put it as a parameter, because currently it's just a debugging technicality.

File: src/models/vae.jl [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+	reconstruct(model::GenerativeModels.VAE, x)
+
+Data reconstruction.
+"""
+reconstruct(model::GenerativeModels.VAE, x) = mean(model.decoder, rand(model.encoder, x))
+
+"""
+	reconstruction_score(model::GenerativeModels.VAE, x)
+
+Anomaly score based on
```
Comment (janfrancu) [inline src/models/vae.jl]: Finish the description.

File: src/models/vae.jl [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+	p = condition(model.decoder, rand(model.encoder, x))
+	-logpdf(p, x)
+end
+"""
+	latent_score(model::GenerativeModels.VAE, x) 
+"""
+function latent_score(model::GenerativeModels.VAE, x) 
+	z = rand(model.encoder, x)
+	-logpdf(model.prior, z)
+end
```
Comment (janfrancu) [inline src/models/vae.jl]: What about the `predict` function abstraction from StatsBase. Is it possible to use it here?

File: scripts/experiments_tabular/vae.jl (hunk 2 above)
Comment (vitskvara) [reply scripts/experiments_tabular/vae.jl:43]: I don't like dividing it by two since then it is truly random anymore - higher values closer to hdim will tend to get selected more often. Also, I would have to redefine the sampling range for the parameter again in `edit_params` which I want to avoid since it would be easy to edit it in once place and not in the other.

File: scripts/experiments_tabular/vae.jl (hunk 3 above)
Comment (vitskvara) [reply scripts/experiments_tabular/vae.jl]: That's a good question. I would do it for 23/5 hours, since then some models will never finish. I will change the input value when the `fit!` function is called in `fit`.

File: scripts/experiments_tabular/vae.jl (hunk 4 above)
Comment (vitskvara) [reply scripts/experiments_tabular/vae.jl:87]: Nope. In `sample_params`, I want to ensure that zdim < hdim, independently on data. Here, I want to ensure that zdim < idim, so that I cant learn identity.

File: src/models/vae.jl (hunk 5 above)
Comment (vitskvara) [reply src/models/vae.jl:29]: I wanted to use it but did not get to it - thanks!

File: src/models/vae.jl (hunk 6 above)
Comment (vitskvara) [reply src/models/vae.jl:57]: Thanks.

File: src/models/vae.jl (hunk 7 above)
Comment (vitskvara) [reply src/models/vae.jl]: Good catch.

File: src/models/vae.jl (hunk 8 above)
Comment (vitskvara) [reply src/models/vae.jl:90]: Yeah it's only used for printing so that there is less output. I can see it being an argument of the function for debugging of model, not really a parameter to be sampled from.

File: src/models/vae.jl (hunk 9 above)
Comment (vitskvara) [reply src/models/vae.jl]: Thanks

File: src/models/vae.jl (hunk 10 above)
Comment (vitskvara) [reply src/models/vae.jl]: Yeah, since there are two different scores, I don't see the necessity for overloading of `predict`.

File: scripts/experiments_tabular/vae.jl (hunk 4 above)
Comment (janfrancu) [reply scripts/experiments_tabular/vae.jl:87]: I see. From afar it looked really similar.

File: scripts/experiments_tabular/vae.jl (hunk 2 above)
Comment (janfrancu) [reply scripts/experiments_tabular/vae.jl:43]: Makes sense.

File: src/models/vae.jl (hunk 10 above)
Comment (janfrancu) [reply src/models/vae.jl]: Ok.

**Review by janfrancu — APPROVED** [review]
